neuralpp/symbolic/sympy_expression.py

```diff
@@ -312,11 +338,63 @@ def from_sympy_function_and_general_arguments(sympy_function: sympy.Basic, argum
             # see test/sympy_test.py: test_sympy_bug()
             sympy_object = sympy_function(*[sympy_argument.sympy_object for sympy_argument in sympy_arguments],
                                           evaluate=False)
-            return SymPyFunctionApplication(sympy_object, type_dict)
+        elif sympy_function == sympy.Piecewise:
+            sympy_object = sympy_piecewise_from_if_then_else(
+                *[sympy_argument.sympy_object for sympy_argument in sympy_arguments])
         else:
             with sympy.evaluate(False):
                 sympy_object = sympy_function(*[sympy_argument.sympy_object for sympy_argument in sympy_arguments])
-                return SymPyFunctionApplication(sympy_object, type_dict)
+        return SymPyFunctionApplication(sympy_object, type_dict)
+
+
+def sympy_piecewise_from_if_then_else(if_: sympy.Basic, then_: sympy.Basic, else_: sympy.Basic) -> sympy.Piecewise:
+    """ In Piecewise, cond comes after clause. """
+    return sympy.Piecewise((then_, if_), (else_, True))
+
+
+def sympy_piecewise_to_if_then_else(piecewise: sympy.Piecewise) -> Tuple[sympy.Basic, sympy.Basic, sympy.Basic]:
+    return piecewise.args[0][1], piecewise.args[0][0], piecewise.args[1][0]
+
+
+def fold_sympy_piecewise(piecewise_args: List[Tuple[sympy.Basic, sympy.Basic]]) -> sympy.Piecewise:
+    """ `fold` any sympy piecewise to 2-entry piecewise. E.g.,
+    Piecewise((s0, c0), (s1, c1), (s2, True)) will be folded into
+    Piecewise((s0, c0), (Piecewise((s1, c1), (s2, True)), True)
+    Note the last condition must be True, otherwise it cannot be transformed into if-then-else
```

Comment:
This line seems out of place, since this function does not regard if-then-else, only piecewise. It seems that for the purposes of this method, that requirement is not needed. Perhaps it is a requirement that needs to be respected in some other part of the code that involves if-then-else?

Reply:
That makes sense. I'll move the check outside this function.